core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -2399,6 +2424,121 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     }
   }
 
+  /**
+   * In this test, we simulate the scene in concurrent jobs using the same
+   * rdd which is marked to do checkpoint:
+   * Job one has already finished the spark job, and start the process of doCheckpoint;
+   * Job two is submitted, and submitMissingTasks is called.
+   * In submitMissingTasks, if taskSerialization is called before doCheckpoint is done,
+   * while part calculates from stage.rdd.partitions is called after doCheckpoint is done,
+   * we may get a ClassCastException when execute the task because of some rdd will do
+   * Partition cast.
+   *
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("SPARK-23053: avoid ClassCastException in concurrent execution with checkpoint") {
```

Comment:
hi @ivoson  -- I'm really sorry but I only just realized that this "test" is really just a repro, and it passes both before and after the actual code changes, since you've replicated the internal logic we're fixing.  As such, I don't think its actually useful as a test case -- perhaps it should get added to the jira as a repro.

I appreciate the work that went into writing this as it helped make the issue clear to me.  I am not sure if there is a good way to test this.  If we can't come up with anything, we should just commit your actual fix, but give me a day or two to think about it ...

Reply:
@squito thanks for reply. I understand this, technically it may not be a UT case, just simulate the scene with exception. I also wonder if there is a good way to test this.

Reply:
hi @ivoson -- I haven't come up with a better way to test this, so I think for now you should

(1) change the PR to *only* include the changes to the DAGScheduler (also undo the `protected[spark]` changes elsewhere)
(2) put this repro on the jira as its a pretty good for showing whats going on.

if we come up with a way to test it, we can always do that later on.

thanks and sorry for the back and forth

Reply:
hi @squito , it's fine. The pr and jira have been updated. Thanks for your patient and review.